// jyframe/modules/app.js
var http = require("http");
var url = require("url");
var regPath = require("./regPath");
module.exports = App;
function App(){
	var midware = this.midware = [];
	var routes = this.routes={
		"get":[],
		"post":[]
	}
	this.server = http.createServer(responseFn);
	function responseFn(req,res){
		var midwareIndex = 0;
		execMiddle();
		function next(){
			midwareIndex++;
			execMiddle()
		}
		function execMiddle(){

			if(midFn=midware[midwareIndex])
			{
				midFn(req,res,next);
			}else{

				var path = url.parse(req.url).pathname;
				var routeFn = null;
				function findItsRoute(routeList){
					for (var i = 0; i < routeList.length; i++) {
						if(routeList[i].url.test(path)){
							req.params = {};
							var params = routeList[i].url.params; 
							params.forEach(function(name,index){
								req.params[name] = RegExp["$"+(index+1)];
							});

							routeFn = routeList[i].fn;
							break;
						}
					};
				}
				switch(req.method){
					case "GET":
						findItsRoute(routes.get);
					break;
					case "POST":
						findItsRoute(routes.post);
				}

				if(routeFn){
					routeFn(req,res);
				}else{
					res.writeHead(404,{"content-type":"text/html;charset=utf-8"});
					res.end("页面不存在");
				}
			}
		}
	}
}
App.prototype.use=function(midFn){
	this.midware.push(midFn);
}
App.prototype.listen = function(port){
	this.port = port;
	this.server.listen(port);
}
App.prototype.get = function(url,getFn){
	this.routes.get.push({url:regPath(url),fn:getFn});
}
App.prototype.post = function(url,postFn){
	this.routes.post.push({url:regPath(url),fn:postFn});
}
